### PR TITLE
Pipeline level exit handler and v2 test 

### PR DESCRIPTION
**Pull request by Linchin**
**Description of your changes:**
Solves #7638. 
* Added pipeline level exit handler
* Changed Makerfile in v2 folder for testing
* Added test case for v2 pipeline level exit handler

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

**Comment by Linchin** [comment]
\retest

File: VERSION [hunk 1]
```diff
@@ -1 +1 @@
-2.0.0-alpha.1
\ No newline at end of file
+2.0.0-alpha.1
```
Comment (chensun) [inline VERSION:1]: revert this change?

File: backend/src/v2/Makefile [hunk 2]
```diff
@@ -18,7 +18,7 @@ SHELL = /bin/bash
 # create an .env file to override dev vars locally
 -include .env
 DEV_IMAGE_PREFIX?=gcr.io/ml-pipeline-test/dev/kfp-
-REPO_ROOT=..
+REPO_ROOT=../../..
```
Comment (chensun) [inline backend/src/v2/Makefile:21]: revert

File: backend/src/v2/Makefile [hunk 3]
```diff
@@ -58,10 +58,13 @@ test-watch: mlmd
 dev: pipeline/v2/hello_world \
 	pipeline/v2/producer_consumer_param \
 	pipeline/test/lightweight_python_functions_v2_pipeline \
-	pipeline/test/lightweight_python_functions_v2_with_outputs
+	pipeline/test/lightweight_python_functions_v2_with_outputs \
+	pipeline/core/loop_static/loop_static_v2\
```
Comment (chensun) [inline backend/src/v2/Makefile:62]: nit: don't mix the loop test with this PR.

File: backend/src/v2/cmd/compiler/main.go [hunk 4]
```diff
@@ -109,8 +109,12 @@ func loadSpec(path string) (*pipelinespec.PipelineJob, error) {
 		return nil, err
 	}
 	spec := &pipelinespec.PipelineSpec{}
-	if err := protojson.Unmarshal(bytes, spec); err != nil {
-		return nil, fmt.Errorf("Failed to parse pipeline spec, error: %s, spec: %v", err, string(bytes))
+	specJson, err := yaml.YAMLToJSON(bytes)
```
Comment (chensun) [inline backend/src/v2/cmd/compiler/main.go:112]: No AI for now, but once we add comments to the yaml, this might fail. We probably need some preprocess here.

File: backend/src/v2/compiler/argocompiler/dag.go [hunk 5]
```diff
@@ -203,6 +204,10 @@ func (c *workflowCompiler) task(name string, task *pipelinespec.PipelineTaskSpec
 			if inputs.iterationIndex == "" {
 				driver.Depends = depends(task.GetDependentTasks())
 			}
+			// Handle exit handler dependency
+			if task.TriggerPolicy.GetStrategy().String() == "ALL_UPSTREAM_TASKS_COMPLETED" {
```
Comment (chensun) [inline backend/src/v2/compiler/argocompiler/dag.go:208]: nit: be consistent on using get method or using field directly.

File: backend/src/v2/compiler/argocompiler/dag.go [hunk 6]
```diff
@@ -494,3 +499,28 @@ func depends(deps []string) string {
...
+func depends_exit_handler(deps []string) string {
+	if len(deps) == 0 {
+		return ""
+	}
+	var builder strings.Builder
+	for index, dep := range deps {
+		if index > 0 {
+			builder.WriteString(" || ")
+		}
+		builder.WriteString(dep)
```
Comment (chensun) [inline backend/src/v2/compiler/argocompiler/dag.go:513]: nit: you could contract these into one line of `builder.WriteString(fmt.Sprintf(...))` or `fmt.Fprintf(&builder, ...)`

[go/gotip/episodes/29](https://goto.google.com/gotip/episodes/29)

File: backend/src/v2/driver/driver.go [hunk 7]
```diff
@@ -436,14 +436,37 @@ func DAG(ctx context.Context, opts Options, mlmd *metadata.Client) (execution *E
 		return execution, fmt.Errorf("ArtifactIterator is not implemented")
 	}
 	isIterator := opts.Task.GetParameterIterator() != nil && opts.IterationIndex < 0
+	// Fan out iterations
```
Comment (chensun) [inline backend/src/v2/driver/driver.go:439]: Are these code mixed from the loop fix?

File: backend/src/v2/Makefile (hunk 2 above)
Comment (Linchin) [reply backend/src/v2/Makefile:21]: I think this needs to be kept if we want to run the Makefile successfully. Probably due to the v2 folder used to be under the root folder.

**Comment by Linchin** [comment]
@chensun I updated the PR and it passed all tests. Would you take another look?

File: VERSION (hunk 1 above)
Comment (Linchin) [reply VERSION]: done

File: backend/src/v2/Makefile (hunk 2 above)
Comment (chensun) [reply backend/src/v2/Makefile:21]: I see. You're right.

Comment (chensun) [reply backend/src/v2/Makefile:21]: I see. You're right.

File: backend/src/v2/Makefile [hunk 8]
```diff
@@ -58,13 +58,15 @@ test-watch: mlmd
...
+	pipeline/test/lightweight_python_functions_v2_with_outputs \
+	pipeline/core/exit_handler/exit_handler_v2 \
+	
 # pipeline/core/condition/condition_v2:
 # pipeline/core/loop_output/loop_output_v2:
-# pipeline/core/loop_static/loop_static_v2:
 # pipeline/core/loop_parameter/loop_parameter_v2:
 # pipeline/core/resource_spec/resource_spec_v2:
 # TODO(v2): migrate v1 samples to v2.
+# pipeline/core/loop_static/loop_static_v2\
```
Comment (chensun) [inline backend/src/v2/Makefile:69]: nit: revert this move?

Comment (chensun) [inline backend/src/v2/Makefile:69]: revert the move?

**Review by chensun — COMMENTED** [review]
/lgtm

Thanks!

File: backend/src/v2/cmd/compiler/main.go (hunk 4 above)
Comment (Linchin) [reply backend/src/v2/cmd/compiler/main.go:112]: Not sure about this, but if the comments are only at the beginning of the yaml, I recall the `yaml.YAMLToJSON()` function just omits it?

File: backend/src/v2/compiler/argocompiler/dag.go (hunk 5 above)
Comment (Linchin) [reply backend/src/v2/compiler/argocompiler/dag.go]: done

File: backend/src/v2/driver/driver.go (hunk 7 above)
Comment (Linchin) [reply backend/src/v2/driver/driver.go]: removed

**Review by chensun — COMMENTED** [review]
/lgtm
/approve

Thanks!
